rclcpp/test/test_init.cpp

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include "rclcpp/utilities.hpp"
+
+TEST(TestInit, is_initialized) {
+
+  EXPECT_FALSE(rclcpp::is_initialized());
+
+  rclcpp::init(0, nullptr);
+
+  EXPECT_TRUE(rclcpp::is_initialized());
+
+  rclcpp::shutdown();
+
+  EXPECT_TRUE(rclcpp::is_initialized());
```

Comment:
What happens if you do:

- `init`
- `shutdown`
- `init`
- `EXPECT_TRUE(is_initialized)`

I would expect `is_initialized()` to be false after shutdown, is that not the expectation? I believe `rcl_ok` should return false if you call `rcl_shutdown` (maybe we're not doing that properly yet).

Reply:
Sorry the previous comment was really two different questions, let me edit it real quick.

Reply:
Yeah, it looks like we're not calling `rcl_shutdown` ever, which until now hasn't had any consequence (none of our middlewares do anything on `rcl_shutdown`/`rmw_shutdown`):

https://github.com/ros2/rclcpp/blob/8f2052d65a8b05e19958db5484f5425c688098b8/rclcpp/src/rclcpp/utilities.cpp#L270-L288

But now it means that `rcl_ok()`'s return value doesn't change when `rclcpp::shutdown()` is used, despite that really being the intention:

- https://github.com/ros2/rcl/blob/d3a734f2661a05fb1ab0746907a3477a02de8dbd/rcl/include/rcl/rcl.h#L137
- https://github.com/ros2/rcl/blob/d3a734f2661a05fb1ab0746907a3477a02de8dbd/rcl/test/rcl/test_rcl.cpp#L140-L143

Reply:
I added the last expectation just to make it clear that the current behaviour is that once initialized, there's no way back.

---

I don't think `init` -> `shutdown` -> `init` works, I tried shutting down in the end of `TestUtilities.init_with_args` and adding a new `init` test after that but an exception is thrown.

---

I'm not sure what the consequences of calling `rcl_shutdown` could be and whether it would be worth doing it on this PR. Whatcha think?

Reply:
I don't know, but I don't want to block this. I think it's clear more work needs to be done though (after merging this).